ui/pages/routes/confirmation-handler.tsx

```diff
@@ -0,0 +1,156 @@
+  const isFullscreen = envType === ENVIRONMENT_TYPE_FULLSCREEN;
+
+  // In test environments (Jest integration tests), run synchronously
+  // In real browser, defer to avoid race conditions with Redux state updates
+  const isTestEnvironment = Boolean(process.env.IN_TEST);
+
+  const showAwaitingSwapScreen = useSelector(getShowAwaitingSwapScreen);
+  const hasSwapsQuotes = useSelector(getHasSwapsQuotes);
+  const hasBridgeQuotes = useSelector(getHasBridgeQuotes);
+  const swapsFetchParams = useSelector(getFetchParams);
+  const pendingApprovals = useSelector(selectPendingApprovalsForNavigation);
+  const hasApprovalFlows = useSelector(getHasApprovalFlows);
+
+  // Read stayOnHomePage from both v5 location.state and v5-compat navState
+  const stayOnHomePage = useMemo(
+    () =>
+      Boolean(location.state?.stayOnHomePage) ||
+      Boolean(navState?.stayOnHomePage),
+    [location.state, navState],
+  );
+
+  const canRedirect = !isNotification && !stayOnHomePage;
+
+  // Flows that *should* navigate in fullscreen, based on E2E specs
+  const hasSnapApproval = pendingApprovals.some((approval) =>
+    snapApprovals.includes(approval.type),
+  );
+
+  // Flows that *should not* navigate in fullscreen, based on E2E specs
+  const hasSmartTransactionStatus = pendingApprovals.some(
+    (approval) =>
+      approval.type === 'smartTransaction:showSmartTransactionStatusPage' &&
+      approval.origin !== 'metamask' &&
+      approval.origin !== 'MetaMask',
+  );
+
+  const skipHandler = useMemo(() => {
+    return (
+      isFullscreen &&
+      (hasSmartTransactionStatus || (!hasSnapApproval && !hasApprovalFlows))
+    );
+  }, [
+    isFullscreen,
+    hasSmartTransactionStatus,
+    hasSnapApproval,
+    hasApprovalFlows,
+  ]);
+
+  // Ported from home.component - checkStatusAndNavigate()
+  useEffect(() => {
+    // Only run when on home/default page (for now)
+    if (pathname !== DEFAULT_ROUTE) {
+      return undefined;
+    }
+
+    if (skipHandler) {
+      return undefined;
+    }
+
+    const checkAndNavigate = () => {
+      if (canRedirect && showAwaitingSwapScreen) {
+        navigate(AWAITING_SWAP_ROUTE);
+      } else if (canRedirect && (hasSwapsQuotes || swapsFetchParams)) {
+        navigate(PREPARE_SWAP_ROUTE);
+      } else if (canRedirect && hasBridgeQuotes) {
+        navigate(CROSS_CHAIN_SWAP_ROUTE + PREPARE_SWAP_ROUTE);
+      } else if (pendingApprovals.length || hasApprovalFlows) {
+        const url = getConfirmationRoute(
+          pendingApprovals?.[0]?.id,
+          pendingApprovals,
+          hasApprovalFlows,
+          '',
+        );
+
+        if (url) {
+          navigate(url, { replace: true });
+        }
+      }
+    };
+
+    // In test environments, run synchronously
+    // In real browser, defer to let Redux state settle after approval resolution
+    if (isTestEnvironment) {
+      checkAndNavigate();
+      return undefined;
+    }
+
+    const timer = setTimeout(checkAndNavigate, 0);
```

Comment:
If we're already running in a `useEffect` which is post render, what impact will this have?